src/Compilers/Core/Portable/IReferenceOrISignature.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.Symbols;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Used to devirtualize Dictionary/HashSet for EqualityComparer{T}.Default
+    /// </summary>
+    internal struct IReferenceOrISignatureEquivalent : IEquatable<IReferenceOrISignatureEquivalent>
+    {
+        private readonly object? _item;
+
+        public IReferenceOrISignatureEquivalent(IReference item)
+        {
+            _item = item;
+        }
+
+        public IReferenceOrISignatureEquivalent(ISignature item)
+        {
+            _item = item;
+        }
+
+        public IReferenceOrISignatureEquivalent(IMethodReference item)
+        {
+            _item = item;
+        }
+
+        public bool Equals(IReferenceOrISignatureEquivalent other)
+        {
+            object? x = _item;
+            object? y = other._item;
+            if (x is null)
+            {
+                return y is null;
+            }
+            else if (ReferenceEquals(x, y))
+            {
+                return true;
+            }
+
+            return Equals(x, y);
+        }
+
+        private new static bool Equals(object? x, object? y)
+        {
+            if (x is null)
+            {
+                return y is null;
+            }
+            else if (ReferenceEquals(x, y))
+            {
+                return true;
+            }
+            else if (x is ISymbolInternal sx && y is ISymbolInternal sy)
+            {
+                return sx.Equals(sy, TypeCompareKind.ConsiderEverything);
+            }
+            else if (x is ISymbolCompareKindComparableInternal cx && y is ISymbolCompareKindComparableInternal cy)
+            {
+                return cx.Equals(cy, TypeCompareKind.ConsiderEverything);
+            }
+            else
+            {
+                return x.Equals(y);
+            }
+        }
+
+        public override bool Equals(object? obj)
+        {
+            return obj is IReferenceOrISignatureEquivalent refOrSig ?
+                Equals(refOrSig) :
+                Equals(_item, obj);
```

Comment:
Was expecting this to be:

```cs
return obj is IReferenceOrISignatureEquivalent refOrSig && Equals(refOrSig);
```

Where will this be validly called with a type other than `IReferenceOrISignatureEqualivent`?

Reply:
Don't think it should ever be validly called. Could change to `throw InvalidOperation`?

Reply:
Usually I just return `false`. It's not wrong it's just not equivalent :smile:

Reply:
Returned false :)